src/App.js

```diff
@@ -1,13 +1,207 @@
+        });
+      })
+      .catch(err => {
+        console.error(`Error! ${err}`);
+      });
+  }
+
+//  sends POST request to add a contact. uses axios to post and calls to the loadingStart
+//  and loadingEnd functions to alert user of network loading event.
+  handleAddContact(attributes) {
+
+    this.loadingStart();
+
+    axios.post('http://localhost:4000/contacts', attributes)
+      .then(resp => {
+
+        this.loadingEnd();
+
+        this.setState({
+          contacts: [...this.state.contacts, resp.data]
+        });
+      })
+      .catch(err => console.log(`Error! ${err}`));
+  }
+
+//  sends DELETE request to remove a specific contact whose ID is used to identify which to DELETE
+//  calls to loadingStart and loadingEnd fucntions to alert user of network loading event.
+  handleDeleteContact(_id) {
+
+    this.loadingStart();
+
+    axios.delete(`http://localhost:4000/contacts/${_id}`)
+      .then(resp => {
+
+        this.loadingEnd();
+
+        const newContacts = this.state.contacts.filter(contact => contact._id !== _id);
+
+        this.setState({
+          contacts: newContacts
+        });
+      })
+      .catch(err => console.log(`Error! ${err}`));
+  }
+
+//  handle functions
+
+//  this handles the search bar change
+  handleSearchBarChange(event) {
+    this.setState({
+      searchText: event.target.value
+    });
+  }
+
+  //  filtering, selecting, getting the selected and clicked function. may have
+  //  redunancy here but if I take one function away the app breaks. Still need
+  //  to play with this section.
+  getFilteredContacts() {
+    //  this removes whites space and converts searchText to lowercase
+    const term = this.state.searchText.trim().toLowerCase();
+
+//  using filter will return a new array of contacts, which will include the
+//  if the function returns true and excluded if the function returns false.
+    return this.state.contacts.filter(contact => {
+      return contact.name.toLowerCase().indexOf(term) >= 0;
+    });
+  }
+
+  getSelectedContacts() {
+
+  //  copied returns from filteredContacts
+    return this.state.contacts.filter((contact) => {
+      return this.state.selectedIds.indexOf(contact._id) >= 0;
+    }
+  );
+  }
+
+  getClickedContacts() {
+
+  //  copied returns from filteredContacts
+
+    return this.state.contacts.filter(contact => {
+      return contact.name.toLowerCase().indexOf(contact._id) >= 0;
+    });
+  }
+
+  // Props id is passed into the function, and the id is pushed
+  // to the selectedContactIds array.
+  handleSelectedContacts(_id) {
+    this.setState({
+      selectedIds: [ ...this.state.selectedIds, _id ]
+    });
+  }
+
+  //  takes the contact out of selectedIds without reset the whole array.
+  //  can't get selected contact to remove from initial contact list onChange
+  //  added to the selected list.
+  handleDeselectedContacts(id) {
+    this.state.selectedIds.splice(this.state.selectedIds.indexOf(id), 1);
```

Comment:
Not the biggest fan of this implementation, I don't think it is quite working either.  I would favor `slice` over `splice`.  Splice actually mutates the array, and `slice` makes a new copy.

I think a better approach here though, would be to simply `filter` over `this.state.selectedIds` and return all ids that were not the `id` passed to the function, and then update the state with that new array.